File: API/1742_Rsi_Trader_V1/CS/RsiTraderV1Strategy.cs
using System;
using System.Collections.Generic;

using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies;

/// <summary>
/// RSI-based trading strategy.
/// Buys when RSI crosses above <see cref="BuyPoint"/> after staying below for two bars.
/// Sells when RSI crosses below <see cref="SellPoint"/> after staying above for two bars.
/// Trades only within the specified time range and supports optional closing of opposite positions.
/// </summary>
public class RsiTraderV1Strategy : Strategy
{
	private readonly StrategyParam<DataType> _candleType;
	private readonly StrategyParam<int> _rsiPeriod;
	private readonly StrategyParam<decimal> _buyPoint;
	private readonly StrategyParam<decimal> _sellPoint;
	private readonly StrategyParam<bool> _closeOnOpposite;
	private readonly StrategyParam<int> _startHour;
	private readonly StrategyParam<int> _endHour;
	private readonly StrategyParam<decimal> _takeProfit;
	private readonly StrategyParam<decimal> _stopLoss;

	private RelativeStrengthIndex _rsi;
	private decimal _prevRsi;
	private decimal _prevPrevRsi;
	private bool _hasPrev;
	private bool _hasPrevPrev;

	/// <summary>
	/// Candle type for strategy calculations.
	/// </summary>
	public DataType CandleType
	{
		get => _candleType.Value;
		set => _candleType.Value = value;
	}

	/// <summary>
	/// RSI period length.
	/// </summary>
	public int RsiPeriod
	{
		get => _rsiPeriod.Value;
		set => _rsiPeriod.Value = value;
	}

	/// <summary>
	/// RSI threshold to open long position.
	/// </summary>
	public decimal BuyPoint
	{
		get => _buyPoint.Value;
		set => _buyPoint.Value = value;
	}

	/// <summary>
	/// RSI threshold to open short position.
	/// </summary>
	public decimal SellPoint
	{
		get => _sellPoint.Value;
		set => _sellPoint.Value = value;
	}

	/// <summary>
	/// Indicates whether to close opposite positions when a new signal appears.
	/// </summary>
	public bool CloseOnOpposite
	{
		get => _closeOnOpposite.Value;
		set => _closeOnOpposite.Value = value;
	}

	/// <summary>
	/// Start hour for trading (0-23).
	/// </summary>
	public int StartHour
	{
		get => _startHour.Value;
		set => _startHour.Value = value;
	}

	/// <summary>
	/// End hour for trading (0-23).
	/// </summary>
	public int EndHour
	{
		get => _endHour.Value;
		set => _endHour.Value = value;
	}

	/// <summary>
	/// Take profit value in price.
	/// </summary>
	public decimal TakeProfit
	{
		get => _takeProfit.Value;
		set => _takeProfit.Value = value;
	}

	/// <summary>
	/// Stop loss value in price.
	/// </summary>
	public decimal StopLoss
	{
		get => _stopLoss.Value;
		set => _stopLoss.Value = value;
	}

	/// <summary>
	/// Initializes a new instance of the strategy.
	/// </summary>
	public RsiTraderV1Strategy()
	{
		_candleType = Param(nameof(CandleType), TimeSpan.FromMinutes(1).TimeFrame())
			.SetDisplay("Candle Type", "Type of candles", "General");

		_rsiPeriod = Param(nameof(RsiPeriod), 14)
			.SetGreaterThanZero()
			.SetDisplay("RSI Period", "Calculation period", "RSI")
			.SetCanOptimize(true)
			.SetOptimize(5, 30, 1);

		_buyPoint = Param(nameof(BuyPoint), 30m)
			.SetDisplay("Buy Threshold", "RSI level for long entry", "RSI")
			.SetCanOptimize(true)
			.SetOptimize(10m, 40m, 5m);

		_sellPoint = Param(nameof(SellPoint), 70m)
			.SetDisplay("Sell Threshold", "RSI level for short entry", "RSI")
			.SetCanOptimize(true)
			.SetOptimize(60m, 90m, 5m);

		_closeOnOpposite = Param(nameof(CloseOnOpposite), true)
			.SetDisplay("Close On Opposite", "Close opposite position on signal", "General");

		_startHour = Param(nameof(StartHour), 0)
			.SetDisplay("Start Hour", "Trading start hour", "Time");

		_endHour = Param(nameof(EndHour), 23)
			.SetDisplay("End Hour", "Trading end hour", "Time");

		_takeProfit = Param(nameof(TakeProfit), 0.005m)
			.SetGreaterThanZero()
			.SetDisplay("Take Profit", "Take profit in price", "Risk")
			.SetCanOptimize(true)
			.SetOptimize(0.001m, 0.02m, 0.001m);

		_stopLoss = Param(nameof(StopLoss), 0.005m)
			.SetGreaterThanZero()
			.SetDisplay("Stop Loss", "Stop loss in price", "Risk")
			.SetCanOptimize(true)
			.SetOptimize(0.001m, 0.02m, 0.001m);
	}

	/// <inheritdoc />
	public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
	{
		return [(Security, CandleType)];
	}

	/// <inheritdoc />
	protected override void OnReseted()
	{
		base.OnReseted();

		_prevRsi = 0m;
		_prevPrevRsi = 0m;
		_hasPrev = false;
		_hasPrevPrev = false;
	}

	/// <inheritdoc />
	protected override void OnStarted(DateTimeOffset time)
	{
		base.OnStarted(time);

		_rsi = new RelativeStrengthIndex { Length = RsiPeriod };

		var subscription = SubscribeCandles(CandleType);
		subscription
			.Bind(_rsi, ProcessCandle)
			.Start();

		var area = CreateChartArea();
		if (area != null)
		{
			DrawCandles(area, subscription);
			DrawIndicator(area, _rsi);
			DrawOwnTrades(area);
		}

		StartProtection(new Unit(TakeProfit, UnitTypes.Price), new Unit(StopLoss, UnitTypes.Price));
	}

	private void ProcessCandle(ICandleMessage candle, decimal rsiValue)
	{
		if (candle.State != CandleStates.Finished)
			return;

		if (!_rsi.IsFormed)
			return;

		var hour = candle.OpenTime.Hour;
		var inTradingTime = hour >= StartHour && hour <= EndHour;
		if (!inTradingTime)
			return;

		if (!_hasPrev)
		{
			_prevRsi = rsiValue;
			_hasPrev = true;
			return;
		}

		if (!_hasPrevPrev)
		{
			_prevPrevRsi = _prevRsi;
			_prevRsi = rsiValue;
			_hasPrevPrev = true;
			return;
		}

		var longSignal = rsiValue > BuyPoint && _prevRsi < BuyPoint && _prevPrevRsi < BuyPoint;
		var shortSignal = rsiValue < SellPoint && _prevRsi > SellPoint && _prevPrevRsi > SellPoint;

		if (longSignal)
		{
			if (CloseOnOpposite && Position < 0)
				BuyMarket(Volume + Math.Abs(Position));
			else if (Position <= 0)
				BuyMarket(Volume);
		}
		else if (shortSignal)
		{
			if (CloseOnOpposite && Position > 0)
				SellMarket(Volume + Math.Abs(Position));
			else if (Position >= 0)
				SellMarket(Volume);
		}

		_prevPrevRsi = _prevRsi;
		_prevRsi = rsiValue;
	}
}

File: API/1742_Rsi_Trader_V1/PY/rsi_trader_v1_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")
clr.AddReference("StockSharp.Algo.Indicators")
clr.AddReference("StockSharp.Algo.Strategies")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import RelativeStrengthIndex
from StockSharp.Algo.Strategies import Strategy


class rsi_trader_v1_strategy(Strategy):
    def __init__(self):
        super(rsi_trader_v1_strategy, self).__init__()
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromHours(4))) \
            .SetDisplay("Candle Type", "Type of candles", "General")
        self._rsi_period = self.Param("RsiPeriod", 14) \
            .SetDisplay("RSI Period", "Calculation period", "RSI")
        self._buy_point = self.Param("BuyPoint", 30.0) \
            .SetDisplay("Buy Threshold", "RSI level for long entry", "RSI")
        self._sell_point = self.Param("SellPoint", 70.0) \
            .SetDisplay("Sell Threshold", "RSI level for short entry", "RSI")
        self._prev_rsi = 0.0
        self._prev_prev_rsi = 0.0
        self._has_prev = False
        self._has_prev_prev = False

    @property
    def candle_type(self):
        return self._candle_type.Value

    @property
    def rsi_period(self):
        return self._rsi_period.Value

    @property
    def buy_point(self):
        return self._buy_point.Value

    @property
    def sell_point(self):
        return self._sell_point.Value

    def OnReseted(self):
        super(rsi_trader_v1_strategy, self).OnReseted()
        self._prev_rsi = 0.0
        self._prev_prev_rsi = 0.0
        self._has_prev = False
        self._has_prev_prev = False

    def OnStarted2(self, time):
        super(rsi_trader_v1_strategy, self).OnStarted2(time)
        rsi = RelativeStrengthIndex()
        rsi.Length = self.rsi_period
        self.SubscribeCandles(self.candle_type).Bind(rsi, self.process_candle).Start()

    def process_candle(self, candle, rsi_value):
        if candle.State != CandleStates.Finished:
            return

        rv = float(rsi_value)

        if not self._has_prev:
            self._prev_rsi = rv
            self._has_prev = True
            return

        if not self._has_prev_prev:
            self._prev_prev_rsi = self._prev_rsi
            self._prev_rsi = rv
            self._has_prev_prev = True
            return

        bp = float(self.buy_point)
        sp = float(self.sell_point)

        long_signal = rv > bp and self._prev_rsi < bp and self._prev_prev_rsi < bp
        short_signal = rv < sp and self._prev_rsi > sp and self._prev_prev_rsi > sp

        if long_signal and self.Position <= 0:
            if self.Position < 0:
                self.BuyMarket()
            self.BuyMarket()
        elif short_signal and self.Position >= 0:
            if self.Position > 0:
                self.SellMarket()
            self.SellMarket()

        self._prev_prev_rsi = self._prev_rsi
        self._prev_rsi = rv

    def CreateClone(self):
        return rsi_trader_v1_strategy()
